# Project : Sales Prediction using Python

**Objective:** Develop a model that predict the Sales based on given parameter.

Dataset consists of 200 rows and 4 column.

We have to predict sales based on following parameter:
1. TV: Expenses on TV advertisement.
2. Radio: Expenses on radio advertisement.
3. Newspaper: Expenses on newspaper advertisement.

**Import libraries**

In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

**Load the data**

In [2]:
df = pd.read_csv(r"D:\Internship\CodSoft\Project 4\advertising.csv")

In [3]:
df.head() # Extrct first 5 rows

TV  Radio  Newspaper  Sales
0  230.1   37.8       69.2   22.1
1   44.5   39.3       45.1   10.4
2   17.2   45.9       69.3   12.0
3  151.5   41.3       58.5   16.5
4  180.8   10.8       58.4   17.9

In [4]:
df.columns    # to see all the columns

Index(['TV', 'Radio', 'Newspaper', 'Sales'], dtype='object')

In [5]:
df.shape   # checking the total rows and columns in the dataset

(200, 4)

In [6]:
df.describe()  # to see the statistical values

TV       Radio   Newspaper       Sales
count  200.000000  200.000000  200.000000  200.000000
mean   147.042500   23.264000   30.554000   15.130500
std     85.854236   14.846809   21.778621    5.283892
min      0.700000    0.000000    0.300000    1.600000
25%     74.375000    9.975000   12.750000   11.000000
50%    149.750000   22.900000   25.750000   16.000000
75%    218.825000   36.525000   45.100000   19.050000
max    296.400000   49.600000  114.000000   27.000000

In [7]:
df.info()  # checking the info of all the columns

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 200 entries, 0 to 199
Data columns (total 4 columns):
 #   Column     Non-Null Count  Dtype  
---  ------     --------------  -----  
 0   TV         200 non-null    float64
 1   Radio      200 non-null    float64
 2   Newspaper  200 non-null    float64
 3   Sales      200 non-null    float64
dtypes: float64(4)
memory usage: 6.4 KB


In [8]:
df.isnull()  # checking the null values

TV  Radio  Newspaper  Sales
0    False  False      False  False
1    False  False      False  False
2    False  False      False  False
3    False  False      False  False
4    False  False      False  False
..     ...    ...        ...    ...
195  False  False      False  False
196  False  False      False  False
197  False  False      False  False
198  False  False      False  False
199  False  False      False  False

[200 rows x 4 columns]

In [9]:
df.isnull().sum()   # checking the total null values in each column

TV           0
Radio        0
Newspaper    0
Sales        0
dtype: int64

In [10]:
x = df.iloc[:, 0:-1].values
y = df.iloc[:, -1].values

In [11]:
print(x)

[[230.1  37.8  69.2]
 [ 44.5  39.3  45.1]
 [ 17.2  45.9  69.3]
 [151.5  41.3  58.5]
 [180.8  10.8  58.4]
 [  8.7  48.9  75. ]
 [ 57.5  32.8  23.5]
 [120.2  19.6  11.6]
 [  8.6   2.1   1. ]
 [199.8   2.6  21.2]
 [ 66.1   5.8  24.2]
 [214.7  24.    4. ]
 [ 23.8  35.1  65.9]
 [ 97.5   7.6   7.2]
 [204.1  32.9  46. ]
 [195.4  47.7  52.9]
 [ 67.8  36.6 114. ]
 [281.4  39.6  55.8]
 [ 69.2  20.5  18.3]
 [147.3  23.9  19.1]
 [218.4  27.7  53.4]
 [237.4   5.1  23.5]
 [ 13.2  15.9  49.6]
 [228.3  16.9  26.2]
 [ 62.3  12.6  18.3]
 [262.9   3.5  19.5]
 [142.9  29.3  12.6]
 [240.1  16.7  22.9]
 [248.8  27.1  22.9]
 [ 70.6  16.   40.8]
 [292.9  28.3  43.2]
 [112.9  17.4  38.6]
 [ 97.2   1.5  30. ]
 [265.6  20.    0.3]
 [ 95.7   1.4   7.4]
 [290.7   4.1   8.5]
 [266.9  43.8   5. ]
 [ 74.7  49.4  45.7]
 [ 43.1  26.7  35.1]
 [228.   37.7  32. ]
 [202.5  22.3  31.6]
 [177.   33.4  38.7]
 [293.6  27.7   1.8]
 [206.9   8.4  26.4]
 [ 25.1  25.7  43.3]
 [175.1  22.5  31.5]
 [ 89.7   9.9  35.7]
 [239.9  41.5

In [12]:
print(y)

[22.1 10.4 12.  16.5 17.9  7.2 11.8 13.2  4.8 15.6 12.6 17.4  9.2 13.7
 19.  22.4 12.5 24.4 11.3 14.6 18.  17.5  5.6 20.5  9.7 17.  15.  20.9
 18.9 10.5 21.4 11.9 13.2 17.4 11.9 17.8 25.4 14.7 10.1 21.5 16.6 17.1
 20.7 17.9  8.5 16.1 10.6 23.2 19.8  9.7 16.4 10.7 22.6 21.2 20.2 23.7
  5.5 13.2 23.8 18.4  8.1 24.2 20.7 14.  16.  11.3 11.  13.4 18.9 22.3
 18.3 12.4  8.8 11.  17.   8.7  6.9 14.2  5.3 11.  11.8 17.3 11.3 13.6
 21.7 20.2 12.  16.  12.9 16.7 14.   7.3 19.4 22.2 11.5 16.9 16.7 20.5
 25.4 17.2 16.7 23.8 19.8 19.7 20.7 15.   7.2 12.   5.3 19.8 18.4 21.8
 17.1 20.9 14.6 12.6 12.2  9.4 15.9  6.6 15.5  7.  16.6 15.2 19.7 10.6
  6.6 11.9 24.7  9.7  1.6 17.7  5.7 19.6 10.8 11.6  9.5 20.8  9.6 20.7
 10.9 19.2 20.1 10.4 12.3 10.3 18.2 25.4 10.9 10.1 16.1 11.6 16.6 16.
 20.6  3.2 15.3 10.1  7.3 12.9 16.4 13.3 19.9 18.  11.9 16.9  8.  17.2
 17.1 20.   8.4 17.5  7.6 16.7 16.5 27.  20.2 16.7 16.8 17.6 15.5 17.2
  8.7 26.2 17.6 22.6 10.3 17.3 20.9  6.7 10.8 11.9  5.9 19.6 17.3  7.6
 14.  1

In [13]:
# splitting into training and testing set

from sklearn.model_selection import train_test_split
x_train, x_test, y_train, y_test = train_test_split(x, y, test_size = 0.70, random_state = 10)

In [14]:
# Applying Linear regression model

from sklearn.linear_model import LinearRegression  
lr = LinearRegression()

In [15]:
lr.fit(x_train, y_train)

LinearRegression()

In [16]:
y_pred = lr.predict(x_test)
y_pred

array([18.94567527, 10.75018474, 19.09746709, 23.93867162, 20.52747377,
       15.11675774, 16.56393106, 20.45565527, 20.79839754, 10.86117477,
       24.29494413,  9.03539661, 10.58777757, 19.81533065, 20.13819992,
       13.48552636, 20.93226156,  8.58732078, 21.42423027, 20.33808618,
       16.87454528,  9.40151398, 22.89598365, 15.02976539, 20.35180656,
        8.3370788 , 10.59744727, 11.55249364, 21.71709085,  8.23685142,
       13.71491021, 20.68897127,  6.62778903,  6.66063507, 20.24592983,
       19.32658014,  9.13565517, 20.62843132, 10.66434665, 16.70508194,
       11.53279837, 13.28487845, 15.48564112, 10.04780937, 21.10209425,
        9.99795748, 16.44856387, 17.76460901,  8.45710135, 13.65920854,
        9.58587419, 16.44882356, 23.93523951, 17.34756217, 17.63918752,
        7.98777246, 24.37117644, 18.44842433, 12.40779825,  9.70324543,
        7.77702274, 17.35524345, 16.72911128, 16.10368024, 18.08224265,
       20.07994685, 17.2479243 , 14.08342025, 13.70103825,  9.34

In [17]:
y_test

array([18.4,  7.2, 18. , 25.5, 22.6, 14.6, 19.9, 23.7, 22.3, 12. , 25.4,
       12.6,  8.7, 20.1, 19.7, 14. , 19.8,  5.3, 21.8, 22.6, 20.6,  1.6,
       24.2, 16. , 19.8,  7. ,  9.5, 10.4, 23.2,  7.6, 12.9, 21.5,  6.9,
        7.3, 17.8, 16.8, 11.9, 20. , 10.6, 16.5, 10.8, 13.2, 15. ,  9.6,
       23.8,  8.8, 19.7, 18.4,  5.5, 12.2,  9.7, 17.9, 23.8, 18. , 17.5,
        8.1, 27. , 18.3, 14. ,  9.7,  6.6, 18.4, 16.9, 17.2, 19.6, 18.9,
       16.9, 14.6, 13.4, 11. , 15.5, 10.1, 14. , 13.3, 17.2, 16.6, 13.6,
        8. ,  9.4, 17.1, 11.8, 17.3, 13.2,  8.4, 19.8, 17. , 21.2, 15. ,
       24.4, 17.3, 10.7, 19. , 11.3, 21.7, 17.6, 10.5, 16.5, 20.5, 25.4,
       20.9, 14.7,  6.7, 10.3, 17.3, 16.1,  5.7, 20.5, 17.6, 11.9, 20.7,
        9.7, 17.7, 11. , 16.4, 16.7, 10.4, 17.1,  9.2, 17.1, 17.4,  5.6,
        7.2, 20.2, 18.2, 11.8, 24.7, 10.1,  3.2, 10.9, 20.7, 11.9,  5.3,
       16.4, 10.8, 12.3, 10.3, 17.9, 19.6,  8.5, 16.7])

In [18]:
print("Mean Absolute Error : " ,metrics.mean_absolute_error(y_test, y_pred))
print("Mean Squared Error : " ,metrics.mean_squared_error(y_test, y_pred))
print("Root Mean Squared Error : " ,np.sqrt(metrics.mean_squared_error(y_test, y_pred)))

<IPython.core.display.Javascript object>

Mean Absolute Error :  1.2782111296680694


<IPython.core.display.Javascript object>

Mean Squared Error :  3.0005649476772036


<IPython.core.display.Javascript object>

Root Mean Squared Error :  1.7322138862384182


In [19]:
print("R squared: {:.2f}".format(lr.score(x,y)*100))

R squared: 89.84


**Interpreting accuracy:**

R square is the coefficient of determination.

It tells us how many points fall on the regression line.

The value of R Square is 89.84, which indicates that 89.84% of the data fit the regression model.

It tells us how many points fall on the regression line.

In [20]:
print("Accuracy Score:" ,lr.score(x_test,y_test)*100)

Accuracy Score: 89.99058322289763


In [21]:
a = [[220, 46, 22]]
y_pred1 = lr.predict(a)
y_pred1

array([21.19329268])